src/Juvix/Syntax/MicroJuvix/Language/Extra.hs

```diff
@@ -113,6 +113,43 @@ mkConcreteType = fmap ConcreteType . go
       e' <- go e
       return (FunctionParameter m i e')
 
+newtype PolyType = PolyType {_unpolyType :: Expression}
+  deriving stock (Eq, Generic)
+
+instance Hashable PolyType
+
+makeLenses ''PolyType
+
+mkPolyType' :: Expression -> PolyType
+mkPolyType' =
+  fromMaybe (error "the given type is not concrete")
+    . mkPolyType
+
+-- mkPolyType removes all named function parameters; currently the assumption in
+-- MicroJuvixToMiniC.hs is that these coincide with type parameters
+mkPolyType :: Expression -> Maybe PolyType
+mkPolyType = fmap PolyType . go
+  where
+    go :: Expression -> Maybe Expression
+    go t = case t of
+      ExpressionApplication (Application l r i) -> do
+        l' <- go l
+        r' <- go r
+        return (ExpressionApplication (Application l' r' i))
+      ExpressionUniverse {} -> return t
+      ExpressionFunction (Function (FunctionParameter m i e) r) | isNothing m -> do
```

Comment:
I think it's better to have:
```
ExpressionFunction (Function (FunctionParameter m i e) r) 
   | isNothing m -> do
        e' <- go e
        r' <- go r
        return (ExpressionFunction (Function (FunctionParameter m i e') r'))
  | otherwise -> go r
```
This way we don't have two patterns that match `ExpressionFunction (Function ...)`

Reply:
Done